### PR TITLE
WRKLDS-577: Migrate templates.sh to template.go

### PR DESCRIPTION
**Pull request by ardaguclu**
This PR migrates oc template test defined in templates.sh to go. Thanks to that, these tests will be used for all cluster variants apart from disconnected.

These tests are skipped for disconnected because template tests require accessing some images.

### CODE REVIEW

**Comment by ardaguclu** [comment]
all failures are infra related;

/retest

**Comment by ardaguclu** [comment]
None of the failures are related to the tests added by this PR

File: test/extended/cli/template.go [hunk 1]
```diff
@@ -16,10 +18,299 @@ var _ = g.Describe("[sig-cli] templates", func() {
...
+		o.Expect(out).To(o.ContainSubstring("ruby-27-centos7"))
+		out, err = oc.Run("process").Args("ruby-helloworld-sample", "-o", "yaml").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("ruby-27-centos7"))
+		out, err = oc.Run("process").Args("ruby-helloworld-sample", "-o", "name").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("ruby-27-centos7"))
+		out, err = oc.Run("describe").Args("templates", "ruby-helloworld-sample").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("ruby-sample-build"))
```
Comment (ingvagabund) [inline test/extended/cli/template.go:81]: Questionable whether `o.Expect(out).To(o.ContainSubstring("BuildConfig"))` should be checked as well to stay aligned with the original `os::cmd::expect_success_and_text 'oc describe templates ruby-helloworld-sample' "BuildConfig.*ruby-sample-build"` line.

File: test/extended/cli/template.go [hunk 2]
```diff
@@ -16,10 +18,299 @@ var _ = g.Describe("[sig-cli] templates", func() {
...
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("frontend-service"))
+		out, err = oc.Run("process").Args("-f", guestbookTemplatePath, "--local", "-l", "app=guestbook", "-o", "yaml").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("app: guestbook"))
+		out, err = oc.Run("process").Args("-f", guestbookTemplatePath, "--local", "-l", "app=guestbook", "-o", "yaml", "--server", "0.0.0.0:1").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("app: guestbook"))
+		err = oc.Run("delete").Args("-f", outputTemplateFile).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
```
Comment (ingvagabund) [inline test/extended/cli/template.go:99]:
```
os::cmd::expect_success 'oc process -f "${guestbook_template}" --local -l app=guestbook -o yaml ADMIN_USERNAME=au ADMIN_PASSWORD=ap REDIS_PASSWORD=rp > "${new}/localtemplate"'
os::cmd::expect_success 'oc process -f "${guestbook_template}" -l app=guestbook -o yaml ADMIN_USERNAME=au ADMIN_PASSWORD=ap REDIS_PASSWORD=rp > "${new}/remotetemplate"'
os::cmd::expect_success 'diff "${new}/localtemplate" "${new}/remotetemplate"'
```
lines were skipped. On purpose?

File: test/extended/cli/template.go [hunk 3]
```diff
@@ -16,10 +18,299 @@ var _ = g.Describe("[sig-cli] templates", func() {
...
+		o.Expect(out).To(o.ContainSubstring("myself"))
+		out, err = oc.Run("process").Args("-p", "MYSQL_USER=myself", "-p", "MYSQL_PASSWORD=my,1%pa=s", "ruby-helloworld-sample").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("my,1%pa=s"))
+		out, err = oc.Run("process").Args("-f", guestbookTemplatePath, fmt.Sprintf("--param-file=%s", guestbookTemplateEnvPath)).Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("root"))
+		o.Expect(out).To(o.ContainSubstring("adminpass"))
+		o.Expect(out).To(o.ContainSubstring("redispass"))
```
Comment (ingvagabund) [inline test/extended/cli/template.go:128]: Missing the same but reading the same input from stdin using cat guestbookTemplatePath. Still worth checking as well. I.e.
```
os::cmd::expect_success_and_text "cat '${guestbook_params}' | oc process -f '${guestbook_template}' --param-file=-" '"root"'
os::cmd::expect_success_and_text "cat '${guestbook_params}' | oc process -f '${guestbook_template}' --param-file=-" '"adminpass"'
os::cmd::expect_success_and_text "cat '${guestbook_params}' | oc process -f '${guestbook_template}' --param-file=-" '"redispass"'
```

File: test/extended/cli/template.go [hunk 4]
```diff
@@ -16,10 +18,299 @@ var _ = g.Describe("[sig-cli] templates", func() {
...
+		out, err = oc.Run("process").Args("-f", templateRequiredParamPath, "--param-file=/dev/null").Output()
+		o.Expect(err).To(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("parameter required_param is required and must be specified"))
+		err = oc.Run("process").Args("-f", guestbookTemplatePath, "--param-file=/dev/null", fmt.Sprintf("--param-file=%s", guestbookTemplateEnvPath)).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		out, err = oc.Run("process").Args("-f", guestbookTemplatePath, "-p", "ABSENT_PARAMETER=absent").Output()
+		o.Expect(err).To(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("unknown parameter name"))
+		err = oc.Run("process").Args("-f", guestbookTemplatePath, "-p", "ABSENT_PARAMETER=absent", "--ignore-unknown-parameters").Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
```
Comment (ingvagabund) [inline test/extended/cli/template.go:158]: Missing
```
os::cmd::expect_failure_and_text "echo 'fo%(o=bar' | oc process -f ${TEST_DATA}/templates/template_required_params.yaml --param-file=-"        'invalid parameter assignment'
os::cmd::expect_failure_and_text "echo 'S P A C E S=test' | oc process -f ${TEST_DATA}/templates/template_required_params.yaml --param-file=-" 'invalid parameter assignment'
```

File: test/extended/cli/template.go [hunk 5]
```diff
@@ -16,10 +18,299 @@ var _ = g.Describe("[sig-cli] templates", func() {
...
+		o.Expect(out).To(o.ContainSubstring(testUser))
+		out, err = oc.AsAdmin().Run("get").Args("user", testUser+"=moreval", "-o", "jsonpath={.metadata.name}").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring(testUser + "=moreval"))
+		out, err = oc.AsAdmin().Run("get").Args("user", testUser+"=moreval2", "-o", "jsonpath={.metadata.name}").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring(testUser + "=moreval2"))
+		out, err = oc.AsAdmin().Run("get").Args("user", testUser+"=moreval3", "-o", "jsonpath={.metadata.name}").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring(testUser + "=moreval3"))
```
Comment (ingvagabund) [inline test/extended/cli/template.go:287]: +1 for checking all the users were created

File: test/extended/cli/template.go [hunk 6]
```diff
@@ -16,10 +18,299 @@ var _ = g.Describe("[sig-cli] templates", func() {
...
+		o.Expect(out).To(o.ContainSubstring("no longer accepts comma-separated list"))
+		out, err = oc.Run("process").Args("-f", templateRequiredParamPath, "--param=required_param=a_b_c_d").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).NotTo(o.ContainSubstring("no longer accepts comma-separated list"))
+		out, err = oc.Run("process").Args("-f", templateRequiredParamPath, "--param=required_param=a,b,c,d").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).NotTo(o.ContainSubstring("no longer accepts comma-separated list"))
+		out, err = oc.Run("process").Args("-f", templateRequiredParamPath, "required_param=a,b=c,d").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).NotTo(o.ContainSubstring("no longer accepts comma-separated list"))
```
Comment (ingvagabund) [inline test/extended/cli/template.go:311]: Missing
```
os::cmd::expect_success_and_text "oc process -f '${required_params}' --param=required_param='`cat ${TEST_DATA}/templates/multiline.txt`'" 'also,with=commas'
```

**Review by ingvagabund — COMMENTED** [review]
Few nits. More like for clarifying.

File: test/extended/cli/template.go (hunk 2 above)
Comment (ardaguclu) [reply test/extended/cli/template.go:99]: First reason is I don't know we support external tools like `diff` or `cat` in origin go tests. Second and more important reason is there are several repetitive tests doing the same thing and these skipped tests are only testing the same thing by combining these external tools.

Comment (ingvagabund) [reply test/extended/cli/template.go:99]: In this case testing `--local` and non-local case produce the same output is still a test case. Worth keeping the case even without the diff. There's `CLI.InputString()` method that can emulate `cat`/`echo` cases to test the `--param-file=-` as well. 

> there are several repetitive tests doing the same thing and these skipped tests are only testing the same thing by combining these external tools

Are you referring to the `--param-file=FILE` and `--param-file=-` cases?

Comment (ardaguclu) [reply test/extended/cli/template.go:99]: > In this case testing `--local` and non-local case produce the same output is still a test case. Worth keeping the case even without the diff. There's `CLI.InputString()` method that can emulate `cat`/`echo` cases to test the `--param-file=-` as well.
> 

--local and --non-local good be a good test case. Thanks.
> > there are several repetitive tests doing the same thing and these skipped tests are only testing the same thing by combining these external tools
> 
> Are you referring to the `--param-file=FILE` and `--param-file=-` cases?

Actually I'm referring to all tests in this file.

**Comment by ardaguclu** [comment]
/retitle WIP: WRKLDS-577: Migrate templates.sh to template.go
until newly added tests are working

**Comment by ardaguclu** [comment]
/retest-required

@ingvagabund it seems that tests are passing. Could you please take a look at one last time?. Thanks.

File: test/extended/cli/template.go [hunk 7]
```diff
@@ -16,10 +19,326 @@ var _ = g.Describe("[sig-cli] templates", func() {
...
+		out, err = oc.Run("process").Args("ruby-helloworld-sample", "-o", "yaml").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("ruby-27-centos7"))
+		out, err = oc.Run("process").Args("ruby-helloworld-sample", "-o", "name").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("ruby-27-centos7"))
+		out, err = oc.Run("describe").Args("templates", "ruby-helloworld-sample").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("BuildConfig"))
+		o.Expect(out).To(o.ContainSubstring("ruby-sample-build"))
```
Comment (soltysh) [inline test/extended/cli/template.go:84]: Next time you can do `o.MatchRegexp('BuildConfig.*ruby-samplebuild')`

**Review by soltysh — APPROVED** [review]
/approve

File: test/extended/cli/template.go (hunk 7 above)
Comment (ardaguclu) [reply test/extended/cli/template.go:84]: Thanks @soltysh, I was looking something like this.

**Pull request merged**